src/power_grid_model/core/serialization.py

```diff
@@ -0,0 +1,319 @@
+# SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+#
+# SPDX-License-Identifier: MPL-2.0
+
+
+"""
+Power grid model (de)serialization
+"""
+
+from abc import ABC, abstractmethod
+from ctypes import byref
+from enum import IntEnum
+from typing import Mapping, Optional, Tuple, Union
+
+import numpy as np
+
+from power_grid_model.core.error_handling import assert_no_error
+from power_grid_model.core.index_integer import IdxC
+from power_grid_model.core.power_grid_core import CharPtr, DeserializerPtr, SerializerPtr, WritableDatasetPtr
+from power_grid_model.core.power_grid_core import power_grid_core as pgc
+from power_grid_model.core.power_grid_dataset import CConstDataset, CWritableDataset
+from power_grid_model.errors import PowerGridSerializationError
+
+
+class SerializationType(IntEnum):
+    """Serialization Format Types"""
+
+    JSON = 0
+    MSGPACK = 1
+
+
+class Deserializer:
+    """
+    Deserializer for the Power grid model
+    """
+
+    _deserializer: DeserializerPtr
+    _dataset_ptr: WritableDatasetPtr
+    _dataset: CWritableDataset
+
+    def __new__(cls, data: Union[str, bytes], serialization_type: SerializationType):
+        instance = super().__new__(cls)
+
+        raw_data = data if isinstance(data, bytes) else data.encode()
+        instance._deserializer = pgc.create_deserializer_from_binary_buffer(
+            raw_data, len(raw_data), serialization_type.value
+        )
+        assert_no_error()
+
+        instance._dataset_ptr = pgc.deserializer_get_dataset(instance._deserializer)
+        assert_no_error()
+
+        instance._dataset = CWritableDataset(instance._dataset_ptr)
+        assert_no_error()
+
+        return instance
+
+    def __del__(self):
+        pgc.destroy_deserializer(self._deserializer)
+
+    def load(self) -> Tuple[Mapping[str, Union[np.ndarray, Mapping[str, np.ndarray]]], str]:
```

Comment:
I think we agreed that we only return the dictionary to the user. Not a tuple. If the user is not sure about the dataset type, the user can call the dataset detection function.

Reply:
the user can only call the dataset detection function if the data type is not empty. what would he have to do when it is empty?

Reply:
This is a very niche corner case, the user have an empty dataset input, and he does not know the dataset type. I would not just for this case create a function with tuple return. Tuple return is in general (I think) confusing.

If the user really has this situation, he/she can always load/unpack the data self and read the `type` key.

Reply:
I would say in most use cases, the user already know the dataset type of a certain file.

Reply:
i exposed the functionality to the user and it raises when it cannot be determined

Reply:
I do not understand. Are you still returning a tuple?

Reply:
nope

Reply:
instead, i exposed the functionality `get_dataset_type(data: Dataset)` to the user